crates/uv-auth/src/middleware.rs

```diff
@@ -40,102 +75,497 @@ impl Middleware for AuthMiddleware {
+            let mut url = request.url().clone();
+            credentials
+                .as_ref()
+                .and_then(|credentials| credentials.username())
+                .and_then(|username| url.set_username(username).ok());
+            credentials
+                .as_ref()
+                .and_then(|credentials| credentials.password())
+                .and_then(|_| url.set_password(Some("****")).ok());
+            url.to_string()
+        } else {
+            request.url().to_string()
+        };
+        trace!("Handling request for {url}");
+
+        // Then check for credentials in (2) the cache
+        let credentials = self.cache().check(request.url(), credentials);
+
+        // Track credentials that we might want to insert into the cache
+        let mut new_credentials = None;
 
-        if let Some(credentials) = GLOBAL_AUTH_STORE.get(&url) {
-            debug!("Adding stored credentials to {url}");
-            request = credentials.authenticated_request(request)
-        } else if GLOBAL_AUTH_STORE.should_attempt_fetch(&url) {
-            if let Some(credentials) = self.nrc.as_ref().and_then(|nrc| {
-                // If we find a matching entry in the netrc file, we can use it
-                url.host_str()
-                    .and_then(|host| nrc.hosts.get(host).or_else(|| nrc.hosts.get("default")))
-            }) {
-                let credentials = Credentials::from(credentials.to_owned());
-                request = credentials.authenticated_request(request);
-                GLOBAL_AUTH_STORE.set(&url, credentials);
+        // If already authenticated (including a password), don't query other services
+        if credentials.is_authenticated() {
+            match credentials {
+                // If we get credentials from the cache, update the request
+                CheckResponse::Cached(credentials) => request = credentials.authenticate(request),
+                // If we get credentials from the request, we should update the cache
+                // but don't need to update the request
+                CheckResponse::Uncached(credentials) => new_credentials = Some(credentials.clone()),
+                CheckResponse::None => unreachable!("No credentials cannot be authenticated"),
+            }
+        // Otherwise, look for complete credentials in:
+        // (3) The netrc file
+        } else if let Some(credentials) = self.netrc.as_ref().and_then(|netrc| {
+            trace!("Checking netrc for credentials for {url}");
+            Credentials::from_netrc(
+                netrc,
+                request.url(),
+                credentials
+                    .get()
+                    .and_then(|credentials| credentials.username()),
+            )
+        }) {
+            debug!("Found credentials in netrc file for {url}");
+            request = credentials.authenticate(request);
+            new_credentials = Some(Arc::new(credentials));
+        // (4) The keyring
```

Comment:
Just confirming: if we have both netrc and keyring enabled, and the URL isn't in the netrc, we subsequently check the keyring, right? (That's my read of the code. Confirming my understanding + that it's intended.)

Reply:
Yep! and that's the same as the behavior on `main`.